[SPARK-34365][AVRO] Add support for positional Catalyst-to-Avro schema matching

xkrogen opened the pull request:
### What changes were proposed in this pull request?
Provide the (configurable) ability to perform Avro-to-Catalyst schema field matching using the position of the fields instead of their names. A new `option` is added for the Avro datasource, `positionalFieldMatching`, which instructs `AvroSerializer`/`AvroDeserializer` to perform positional field matching instead of matching by name.

### Why are the changes needed?
This by-name matching is somewhat recent; prior to PR #24635, at least on the write path, schemas were matched by positionally ("structural" comparison). While by-name is better behavior as a default, it will be better to make this configurable by a user. Even at the time that PR #24635 was handled, there was [interest in making this behavior configurable](https://github.com/apache/spark/pull/24635#issuecomment-494205251), but it appears it went unaddressed.

There is precedence for configurability of this behavior as seen in PR #29737, which added this support for ORC. Besides this precedence, the behavior of Hive is to perform matching positionally ([ref](https://cwiki.apache.org/confluence/display/Hive/AvroSerDe#AvroSerDe-WritingtablestoAvrofiles)), so this is behavior that Hadoop/Hive ecosystem users are familiar with.

### Does this PR introduce _any_ user-facing change?
Yes, a new option is provided for the Avro datasource, `positionalFieldMatching`, which provides compatibility with Hive and pre-3.0.0 Spark behavior.

### How was this patch tested?
New unit tests are added within `AvroSuite`, `AvroSchemaHelperSuite`, and `AvroSerdeSuite`; and most of the existing tests within `AvroSerdeSuite` are adapted to perform the same test using by-name and positional matching to ensure feature parity.

xkrogen committed:
[SPARK-34365][AVRO] Add support for schemaRenaming to perform positional Catalyst-to-Avro schema matching and unify AvroSerializer vs. AvroDeserializer field matching behavior.

xkrogen commented:
One area I am particularly open to feedback on is using a datasource option vs. a SQL config. Internally we have been using an option for a number of years and have found it useful for being able to control the behavior on a per-schema basis, but PR #29737 creates some precedence for a SQL config.

xkrogen commented:
cc folks who have worked on either #29737 or #24635: @cloud-fan @dongjoon-hyun @dbtsai @viirya @gatorsmile @peter-toth

SparkQA commented:
**[Test build #134936 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/134936/testReport)** for PR 31490 at commit [`7e687cb`](https://github.com/apache/spark/commit/7e687cbc17fb0bfa7c532696aa9a3b6abd1497f8).

SparkQA commented:
**[Test build #134936 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/134936/testReport)** for PR 31490 at commit [`7e687cb`](https://github.com/apache/spark/commit/7e687cbc17fb0bfa7c532696aa9a3b6abd1497f8).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

SparkQA commented:
Kubernetes integration test starting
URL: https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder-K8s/39519/

SparkQA commented:
Kubernetes integration test status failure
URL: https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder-K8s/39519/

AmplabJenkins commented:
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/134936/

AmplabJenkins commented:
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/39519/

xkrogen commented:
Rebase on top of recent changes in master.
@MaxGekk or @cloud-fan, I see you have some changes in this area recently, are either of you interested in taking a look?

SparkQA commented:
**[Test build #136478 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/136478/testReport)** for PR 31490 at commit [`5074a2a`](https://github.com/apache/spark/commit/5074a2abee870b87dba9ab0f595f44c76194d2ef).

SparkQA commented:
**[Test build #136478 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/136478/testReport)** for PR 31490 at commit [`5074a2a`](https://github.com/apache/spark/commit/5074a2abee870b87dba9ab0f595f44c76194d2ef).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/136478/

SparkQA commented:
Kubernetes integration test starting
URL: https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder-K8s/41062/

SparkQA commented:
Kubernetes integration test status failure
URL: https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder-K8s/41062/

AmplabJenkins commented:
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/41062/

xkrogen commented:
@mridulm do you have any interest in helping to review this?

HyukjinKwon commented:
cc @gengliangwang too

xkrogen commented:
Rebased on top of `master` to resolve conflicts.
Gentle ping to @cloud-fan @gengliangwang if you're interested in helping to review.

SparkQA commented:
**[Test build #138942 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/138942/testReport)** for PR 31490 at commit [`1622032`](https://github.com/apache/spark/commit/1622032c0167c75e8e028c5b65a676ddeb8ef53b).

SparkQA commented:
**[Test build #138942 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/138942/testReport)** for PR 31490 at commit [`1622032`](https://github.com/apache/spark/commit/1622032c0167c75e8e028c5b65a676ddeb8ef53b).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

SparkQA commented:
Kubernetes integration test starting
URL: https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder-K8s/43462/

AmplabJenkins commented:
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/138942/

SparkQA commented:
Kubernetes integration test status success
URL: https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder-K8s/43462/

AmplabJenkins commented:
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/43462/

xkrogen commented:
Friendly ping to @mridulm @cloud-fan @gengliangwang for any help with reviews.

mridulm commented on external/avro/src/main/scala/org/apache/spark/sql/avro/AvroUtils.scala:
```diff
@@ -203,34 +203,50 @@ private[sql] object AvroUtils extends Logging {
+    val avroFields = avroSchema.getFields.asScala
+    val catalystMatches = catalystType.zipWithIndex.map { case (catalystField, catalystPos) =>
+      val avroField = if (positionalFieldMatch) {
+        avroFields.drop(catalystPos).headOption
```
`applyOrElse` instead ?

xkrogen replied:
Yes, great call! Thanks for teaching me, this is exactly what I was looking for.

xkrogen replied:
Actually, what I wanted was `lift`, since I also need to convert to `Option` along the way.

mridulm replied:
`lift` sounded expensive here

mridulm commented on external/avro/src/main/scala/org/apache/spark/sql/avro/AvroUtils.scala:
```diff
@@ -203,34 +203,50 @@ private[sql] object AvroUtils extends Logging {
-      )
+    val avroFields = avroSchema.getFields.asScala
+    val catalystMatches = catalystType.zipWithIndex.map { case (catalystField, catalystPos) =>
+      val avroField = if (positionalFieldMatch) {
```
Dont we want to check for type compatibility here ?

xkrogen replied:
This is just deciding which fields should be attempted for matching. The type compatibility is handled after this, in `AvroDeserializer#newWriter` and `AvroSerializer#newConverter`.

mridulm replied:
Thanks for clarifying !

mridulm reviewed (commented):
Took a pass through this.
Will let @HyukjinKwon, @cloud-fan or @gengliangwang review more.

gengliangwang commented on external/avro/src/main/scala/org/apache/spark/sql/avro/AvroOptions.scala:
```diff
@@ -67,6 +67,13 @@ private[sql] class AvroOptions(
+   * name. This allows for a structurally equivalent Catalyst schema to be used with an Avro schema
+   * whose field names do not match. Defaults to false.
+   */
+  val schemaRenaming: Boolean = parameters.get("schemaRenaming").exists(_.toBoolean)
```
shall we make it "writeByPosition"

xkrogen replied:
This is used on the read side as well, so I don't think "writeByPosition" is correct. How about "positionalFieldMatching" or "matchFieldsByPosition"?

xkrogen replied:
Updated to `positionalFieldMatching` in the latest diff.

gengliangwang commented:
@xkrogen Do you mind holding this one until https://github.com/apache/spark/pull/32969 is merged?  I believe there will be conflicts.

xkrogen commented on external/avro/src/main/scala/org/apache/spark/sql/avro/AvroOptions.scala:
```diff
@@ -67,6 +67,13 @@ private[sql] class AvroOptions(
   }
 
+  /**
+   * Iff true, perform Catalyst-to-Avro schema matching based on field position instead of field
```
Iff is intentional here. [It means "if and only if"](https://en.wikipedia.org/wiki/If_and_only_if) and is [pretty widely used in the Spark codebase](https://github.com/apache/spark/search?q=repo%3Aapache%2Fspark+iff).

xkrogen commented:
@gengliangwang thanks for looking! Yes I'd be happy to wait for #32969 as I think that is higher priority and looks close to merging.

gengliangwang commented on external/avro/src/main/scala/org/apache/spark/sql/avro/AvroUtils.scala:
```diff
@@ -203,34 +203,50 @@ private[sql] object AvroUtils extends Logging {
+      val avroField = if (positionalFieldMatch) {
+        avroFields.drop(catalystPos).headOption
+      } else {
+        avroFields.filter(f => SQLConf.get.resolver(f.name(), catalystField.name)).toSeq match {
```
Replying the comment in https://github.com/apache/spark/pull/32969#issuecomment-866130603:
The above line still has perf regression if we match by names.

xkrogen replied:
Yes you're right, I spoke too quickly :) Updated the linked comment, PTAL.

gengliangwang commented:
@xkrogen https://github.com/apache/spark/pull/32969 is merged. You can resolve the conflict and continue the work now. Thanks for the work!

SparkQA commented:
**[Test build #140227 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/140227/testReport)** for PR 31490 at commit [`a83c8c7`](https://github.com/apache/spark/commit/a83c8c750eac6202f8684aa9d744e7627295d91f).

xkrogen commented:
Just pushed up a new version which is pretty much a complete re-implementation on top of #32969.
This still consolidates more of the field matching behavior between `AvroSerializer` and `AvroDeserializer`, as in the original PR, but it leverages the nature of `AvroSchemaHelper` to serve the results to the caller in what I think is a more clean way. Instead of passing back a `Seq[(Option, Option)]` which the caller then pattern-matches on to handle the various cases, the `AvroSchemaHelper` performs all of the matching up front and then provides three methods to access all of the fields matching various conditions (present in both schemas, only Avro, or only Catalyst). I think this makes the logic in the callers more simple and clear.

I plan to also update benchmarks tomorrow to ensure that there is no regression on top of #32969.

SparkQA commented:
**[Test build #140227 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/140227/testReport)** for PR 31490 at commit [`a83c8c7`](https://github.com/apache/spark/commit/a83c8c750eac6202f8684aa9d744e7627295d91f).
 * This patch **fails Scala style tests**.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:
  * `  class AvroSchemaHelper(`

AmplabJenkins commented:
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/140227/

SparkQA commented:
Kubernetes integration test starting
URL: https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder-K8s/44755/

SparkQA commented:
Kubernetes integration test status failure
URL: https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder-K8s/44755/

AmplabJenkins commented:
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/44755/

xkrogen committed:
Scalastyle fixes

SparkQA commented:
**[Test build #140285 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/140285/testReport)** for PR 31490 at commit [`79f7b96`](https://github.com/apache/spark/commit/79f7b96cf498083653555e6e70c11b481841c45b).

xkrogen commented:
Pushed up a style fix. Just remembered that we added the ability to run benchmarks via GitHub Actions, trying that out now.

SparkQA commented:
**[Test build #140285 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/140285/testReport)** for PR 31490 at commit [`79f7b96`](https://github.com/apache/spark/commit/79f7b96cf498083653555e6e70c11b481841c45b).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/140285/

SparkQA commented:
Kubernetes integration test unable to build dist.

exiting with code: 1
URL: https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder-K8s/44817/

AmplabJenkins commented:
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/44817/

gengliangwang commented on external/avro/src/main/scala/org/apache/spark/sql/avro/AvroDeserializer.scala:
```diff
@@ -335,36 +336,32 @@ private[sql] class AvroDeserializer(
+    val avroSchemaHelper =
+      new AvroUtils.AvroSchemaHelper(avroType, catalystType, avroPath, positionalFieldMatch)
+
+    avroSchemaHelper.getCatalystFieldsWithoutMatch.filterNot(_.nullable) match {
```
Why do we need this step which goes over the two schemas? We can just check the length of the two schemas is the same and every Avro field has a corresponding catalyst field in the loop below.

xkrogen replied:
On the `AvroSerializer`, we could just check the schema lengths, but then when there is a mismatch, you don't get a very helpful error message: "Expected 7 fields but got 6" is not nearly as helpful as "Avro schema is missing field 'foo'". I think it's worth the extra complexity here to have more useful error messages, and I think this argument will be strengthened by the `AvroDeserializer` discussion below.

On the deserializer side, checking the length is not sufficient. This is because we actually allow Avro fields which don't have a matched Catalyst field (note that we don't bother calling `avroSchemaHelper.getAvroFieldsWithoutMatch` in the deserializer), treating this as projection. Similarly, we allow Catalyst fields which don't have a matching Avro field, as long as the Catalyst field is nullable, hence the `filterNot(_.nullable)` call. Neither of these things would be possible if we just checked the schema length. Note that even before this PR, we only perform a check on the schema lengths in the serializer, not the deserializer.

Given that it's actually necessary for us to implement this logic on the deserializer side, I think it makes sense to also leverage it on the serializer side to provide better error messages.

gengliangwang replied:
Shall we have this improvement in another PR? And we can have test cases of incompatible schemas for both by-name and by-position matching. 
Currently I find this PR a bit complicated and there is not test cases for this.
Supporting the position matching in this one is enough.

xkrogen replied:
Sure, reverted, thanks for the input!

gengliangwang commented on external/avro/src/main/scala/org/apache/spark/sql/avro/AvroSerializer.scala:
```diff
@@ -240,29 +240,30 @@ private[sql] class AvroSerializer(
-      throw new IncompatibleSchemaException(
-        s"Avro $avroPathStr schema length (${avroFields.size}) doesn't match " +
-        s"SQL ${toFieldStr(catalystPath)} schema length (${catalystStruct.length})")
+    avroSchemaHelper.getAvroFieldsWithoutMatch.toSeq match {
```
Same here.

xkrogen replied:
Responding on the `AvroDeserializer` comment to keep discussion consolidated.

xkrogen committed:
check in the benchmarks based off of the current master

xkrogen committed:
include benchmarks based off of the current diff

xkrogen commented:
I ran all 4 of the Avro benchmarks (read/write on JDK 8/11) against the current diff. I noticed that the JDK11 benchmarks weren't updated when #32969 went in, and also, those were generated using different hardware since they didn't use GitHub Actions. For this reason, I also generated all of the benchmark output files using GitHub Actions against the current `master` branch, applied those, and then applied the benchmarks for this branch on top of those. You can view the diff in this commit: https://github.com/apache/spark/pull/31490/commits/a0b4ecd336b63b2bbad3ff80a30249cba590f053
(numbers look essentially the same, well within the variance I think we expect)

SparkQA commented:
**[Test build #140349 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/140349/testReport)** for PR 31490 at commit [`a0b4ecd`](https://github.com/apache/spark/commit/a0b4ecd336b63b2bbad3ff80a30249cba590f053).

SparkQA commented:
**[Test build #140349 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/140349/testReport)** for PR 31490 at commit [`a0b4ecd`](https://github.com/apache/spark/commit/a0b4ecd336b63b2bbad3ff80a30249cba590f053).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

SparkQA commented:
Kubernetes integration test starting
URL: https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder-K8s/44880/

AmplabJenkins commented:
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/140349/

AmplabJenkins commented:
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/44880/

gengliangwang commented:
@xkrogen Thanks for the work. Some high-level suggestions:
1. we should document the new option in sql-data-source-avro.md
2. We don't need to update the benchmark results in the PR.
3. Let's have another PR for the error message improvement of incompatible schemas with test cases.

xkrogen committed:
revert benchmark updates

xkrogen committed:
Revert changes that aren't strictly necessary for the positional field matching functionality.

xkrogen committed:
Update sql-data-sources-avro to document the new option.

xkrogen commented:
Thanks for the input @gengliangwang ! 
1. Done, thanks for the pointer!
2. Reverted
3. Sure, I have reverted some of the changes to keep this PR more narrow in scope to just adding the support for positional matching. Created SPARK-35918 to track this follow-up.

gengliangwang commented on external/avro/src/test/scala/org/apache/spark/sql/avro/AvroSerdeSuite.scala:
```diff
@@ -144,6 +173,43 @@ object AvroSerdeSuite {
+      case BY_POSITION => true
+    }
+    serdeType match {
+      case SERIALIZE => Left(new AvroSerializer(catalystSchema, avroSchema, false, positional,
```
QQ: why  using Left/Right here?

xkrogen replied:
This is a way for `createSerde` to return either `AvroSerializer` or `AvroDeserializer`, depending on the input value (`SERIALIZE` vs `DESERIALIZE`).

I realized a cleaner way to do this is to pass the serializer/deserializer as a factory object which is properly typed. Then the caller doesn't have to extract the proper side of the `Either`. I updated the PR.

SparkQA commented:
**[Test build #140391 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/140391/testReport)** for PR 31490 at commit [`63bed24`](https://github.com/apache/spark/commit/63bed241a1c1964ddf2b4b54741b750145752243).

xkrogen committed:
Get rid of the confusing Either in AvroSerdeSuite, replace the enum with a factory object

SparkQA commented:
**[Test build #140391 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/140391/testReport)** for PR 31490 at commit [`63bed24`](https://github.com/apache/spark/commit/63bed241a1c1964ddf2b4b54741b750145752243).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/140391/

SparkQA commented:
**[Test build #140398 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/140398/testReport)** for PR 31490 at commit [`c7eea8a`](https://github.com/apache/spark/commit/c7eea8a3eb7495f1908c0b851981b22bc5e024c2).

SparkQA commented:
**[Test build #140398 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/140398/testReport)** for PR 31490 at commit [`c7eea8a`](https://github.com/apache/spark/commit/c7eea8a3eb7495f1908c0b851981b22bc5e024c2).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/140398/

SparkQA commented:
Kubernetes integration test starting
URL: https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder-K8s/44915/

SparkQA commented:
Kubernetes integration test status success
URL: https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder-K8s/44915/

AmplabJenkins commented:
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/44915/

gengliangwang commented:
Thanks, merging to master

xkrogen commented:
Thanks @gengliangwang ! Your feedback was much appreciated.
Thanks also @mridulm for the early review.
